--- pango-1.13.1/pango/opentype/ftglue.c.orig	2006-05-19 01:43:58.000000000 -0700
+++ pango-1.13.1/pango/opentype/ftglue.c	2006-05-19 01:44:06.000000000 -0700
@@ -281,7 +281,7 @@
       if ( tag == the_tag )
       {
         LOG(( "TrueType table (start: %ld) (size: %ld)\n", start, size ));
-        error = _hb_ftglue_stream_seek( stream, offset+start );
+        error = _hb_ftglue_stream_seek( stream, start );
         goto FoundIt;
       }
     }
